src/pkcs15init/pkcs15-lib.c

```diff
@@ -1664,6 +1665,8 @@ sc_pkcs15init_generate_key(struct sc_pkcs15_card *p15card, struct sc_profile *pr
 
 err:
 	sc_pkcs15_free_object(object);
+	if (r < 0 && alg_id_free)
+		free(pubkey->alg_id);
```

Comment:
I think the `pubkey->alg_id` should be freed regardless the return of this function as it is pointer to the `pubkey_args`, which is local to this function. It is passed to the `sc_pkcs15init_store_public_key()`, but if it needs it for something, it should copy it out similarly as with the ec params discussed in #2790.

And given that the `pubkey_args` is correctly zeroized in the beginning of this function, I think the `alg_id_free` is not needed either. It will be allocated either here or in the generate_key provided by the card driver and there is no card-specific free after this ...